app/gui/src/controller/searcher/input.rs

```diff
@@ -0,0 +1,401 @@
+            ""
+        }
+    }
+
+    /// Return the accessor chain being the context of the edited name, i.e. the preceding fully
+    /// qualified name parts or this argument.
+    pub fn context(&self) -> Option<ast::opr::Chain> {
+        self.edited_ast.edited_accessor_chain.clone().map(|mut chain| {
+            chain.ast.args.pop();
+            chain.ast
+        })
+    }
+
+    /// The range of the text which shall be replaced by applied suggestion.
+    pub fn replaced_range(&self) -> text::Range<text::Byte> {
+        match &self.edited_ast.edited_name {
+            Some(name) => name.range,
+            None => (self.cursor_position..self.cursor_position).into(),
+        }
+    }
+
+    /// The input as AST node. Returns [`None`] if the input is not a valid AST.
+    pub fn ast(&self) -> Option<&Ast> {
+        match &self.ast {
+            InputAst::Line(ast) => Some(&ast.elem),
+            InputAst::Invalid(_) => None,
+        }
+    }
+}
+
+
+// === Inserting Suggestion ===
+
+/// The description of the results of inserting suggestion.
+#[derive(Clone, Debug)]
+pub struct InsertedSuggestion {
+    /// An entire input after inserting the suggestion.
+    pub new_input:     String,
+    /// The replaced range in the old input.
+    pub replaced:      text::Range<text::Byte>,
+    /// The range of the inserted code in the new input. It does not contain any additional spaces
+    /// (in contrary to [`inserted_text`] field.
+    pub inserted_code: text::Range<text::Byte>,
+    /// The range of the entire inserted text in the new input. It contains code and all additional
+    /// spaces.
+    pub inserted_text: text::Range<text::Byte>,
+}
+
+
+impl InsertedSuggestion {
+    /// The text change resulting from this insertion.
+    pub fn input_change(&self) -> text::Change<text::Byte, String> {
+        let to_insert = self.new_input[self.inserted_text].to_owned();
+        text::Change { range: self.replaced, text: to_insert }
+    }
+}
+
+impl Input {
+    /// Return an information about input change after inserting given suggestion.
+    pub fn after_inserting_suggestion(
+        &self,
+        suggestion: &action::Suggestion,
+        has_this: bool,
+    ) -> FallibleResult<InsertedSuggestion> {
+        let replaced = self.replaced_range();
+        let generate_this = self.context().is_none() && !has_this;
+        let code_to_insert = suggestion.code_to_insert(generate_this);
+        debug!("Code to insert: \"{code_to_insert}\"");
+        let end_of_inserted_code = replaced.start + text::Bytes(code_to_insert.len());
+        let end_of_inserted_text = end_of_inserted_code + text::Bytes(1);
+        let mut new_input = self.ast.to_string();
+        let raw_range = replaced.start.value..replaced.end.value;
+        let range_start = raw_range.start;
+        let range_end = raw_range.end;
+        Self::ensure_on_char_boundary(&new_input, range_start)?;
+        Self::ensure_on_char_boundary(&new_input, range_end)?;
```

Comment:
I saw rare panics when one of the indices was not on char boundary inside the input, which was weird. This code helps to mitigate the issue if it ever happens again.